File: app/src/main/java/com/zionhuang/music/ui/adapters/SearchResultAdapter.java
package com.zionhuang.music.ui.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.paging.PagingDataAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.google.api.services.youtube.model.SearchResult;
import com.zionhuang.music.databinding.ItemSearchResultBinding;

import java.util.Objects;

public class SearchResultAdapter extends PagingDataAdapter<SearchResult, SearchResultAdapter.ViewHolder> {
    private static final String TAG = "SearchResultAdapter";

    public SearchResultAdapter() {
        super(new ItemComparator());
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ItemSearchResultBinding videoBinding = ItemSearchResultBinding.inflate(inflater, parent, false);
        return new ViewHolder(videoBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    @Override
    public int getItemViewType(int position) {
        switch (Objects.requireNonNull(getItem(position)).getId().getKind()) {
            case "youtube#video":
                return 0;
            case "youtube#channel":
                return 1;
            case "youtube#playlist":
                return 2;
        }
        return -1;
    }

    public SearchResult getItemByPosition(int position) {
        return getItem(position);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private ItemSearchResultBinding binding;

        public ViewHolder(ItemSearchResultBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(SearchResult item) {
            binding.setItem(item);
        }
    }

    static class ItemComparator extends DiffUtil.ItemCallback<SearchResult> {
        @Override
        public boolean areItemsTheSame(@NonNull SearchResult oldItem, @NonNull SearchResult newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull SearchResult oldItem, @NonNull SearchResult newItem) {
            return oldItem.getEtag().equals(newItem.getEtag());
        }
    }
}

File: app/src/main/java/com/zionhuang/music/ui/adapters/SearchResultAdapter.kt
package com.zionhuang.music.ui.adapters

import android.view.ViewGroup
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import com.google.api.services.youtube.model.SearchResult
import com.zionhuang.music.R
import com.zionhuang.music.extensions.inflateWithBinding
import com.zionhuang.music.ui.viewholders.SearchChannelViewHolder
import com.zionhuang.music.ui.viewholders.SearchPlaylistViewHolder
import com.zionhuang.music.ui.viewholders.SearchStreamViewHolder
import com.zionhuang.music.ui.viewholders.base.SearchViewHolder

class SearchResultAdapter : PagingDataAdapter<SearchResult, SearchViewHolder>(SearchResultItemComparator()) {
    companion object {
        const val ITEM_VIDEO = 0
        const val ITEM_CHANNEL = 1
        const val ITEM_PLAYLIST = 2
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SearchViewHolder = when (viewType) {
        ITEM_VIDEO -> SearchStreamViewHolder(parent.inflateWithBinding(R.layout.item_search_stream))
        ITEM_CHANNEL -> SearchChannelViewHolder(parent.inflateWithBinding(R.layout.item_search_channel))
        ITEM_PLAYLIST -> SearchPlaylistViewHolder(parent.inflateWithBinding(R.layout.item_search_playlist))
        else -> throw IllegalArgumentException("Unexpected item type.")
    }

    override fun onBindViewHolder(holder: SearchViewHolder, position: Int) {
        when (holder) {
            is SearchStreamViewHolder -> holder.bind(getItem(position)!!)
            is SearchChannelViewHolder -> holder.bind(getItem(position)!!)
            is SearchPlaylistViewHolder -> holder.bind(getItem(position)!!)
        }
    }

    override fun getItemViewType(position: Int): Int = when (getItem(position)?.id?.kind) {
        "youtube#video" -> ITEM_VIDEO
        "youtube#channel" -> ITEM_CHANNEL
        "youtube#playlist" -> ITEM_PLAYLIST
        else -> -1
    }

    fun getItemByPosition(position: Int) = getItem(position)

    class SearchResultItemComparator : DiffUtil.ItemCallback<SearchResult>() {
        override fun areItemsTheSame(oldItem: SearchResult, newItem: SearchResult): Boolean = oldItem.id == newItem.id
        override fun areContentsTheSame(oldItem: SearchResult, newItem: SearchResult): Boolean = oldItem.etag == newItem.etag
    }
}